flocker/control/test/test_model_change.py

```diff
@@ -0,0 +1,418 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+Whenever the configuration model changes we need to write code to
+upgrade the on-disk format from previous releases. This module will
+automatically detect such changes by failing a test, ensuring that upgrade
+code is always implemented when necessary.
+"""
+
+from json import loads, dumps
+from subprocess import check_output
+
+from pyrsistent import (
+    PRecord, PClass, CheckedPSet, CheckedPVector, CheckedPMap, field,
+)
+
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.python.filepath import FilePath
+from twisted.python.reflect import qual as fqpn
+
+from .._model import Deployment
+
+PERSISTED_MODEL = FilePath(__file__).sibling(b"persisted_model.json")
+
+# The class at the root of the configuration tree. This may need to be
+# changed if the configuration's root class changes.
+ROOT_CLASS = Deployment
+
+
+def generate_model(root_class=ROOT_CLASS):
+    """
+    Generate a data-structure that represents the current configuration
+    model.
+
+    Changes to output may require regenerating the persisted version
+    on-disk.
+
+    This should switch to Pyrsistent's introspection API once it exists:
+    https://github.com/tobgu/pyrsistent/issues/47
+    """
+    classes_result = {}
+    result = {u"root": fqpn(root_class),
+              u"classes": classes_result}
+    classes = {root_class}
+    while classes:
+        klass = classes.pop()
+        klass_name = fqpn(klass)
+        if klass_name in classes_result:
+            continue
+        record = None
+        if issubclass(klass, (PRecord, PClass)):
+            if issubclass(klass, PRecord):
+                attr_name = "_precord_fields"
+            else:
+                attr_name = "_pclass_fields"
+            record = {u"category": u"record",
+                      u"fields": {}}
+            for name, field_info in getattr(klass, attr_name).items():
+                record[u"fields"][name] = list(
+                    fqpn(cls) for cls in field_info.type)
```

Comment:
This is why you like tox, isn't it?  Because it sets PYTHONHASHSEED=random every time?

A small number of simple tests should quickly show whether there's a problem or not.  That seems worth doing now rather than possibly making this someone else's problem in a few days.